lms/services/async_oauth_http.py

```diff
@@ -0,0 +1,69 @@
+import asyncio
+from typing import List
+
+import aiohttp
+
+from lms.services import ExternalRequestError
+
+
+class AsyncOAuthHTTPService:
+    def __init__(self, oauth2_token_service):
+        self._oauth2_token_service = oauth2_token_service
+
+    def request(
+        self, method, urls, concurrency=10, headers=None, timeout=10
+    ) -> List[aiohttp.ClientResponse]:
+        """
+        Sends async requests to all urls.
+
+        :raise OAuth2TokenError: if we don't have an access token for the user
+        :raise ExternalRequestError: if something goes wrong with the HTTP
+            request
+        """
+
+        headers = headers or {}
+
+        access_token = self._oauth2_token_service.get().access_token
+        headers["Authorization"] = f"Bearer {access_token}"
+        return asyncio.run(
+            _prepare_requests(
+                method, urls, concurrency, timeout=timeout, headers=headers
+            )
+        )
+
+
+async def _async_request(semaphore, aio_session, method, url, **kwargs):
+    try:
+        async with semaphore:
+            async with aio_session.request(method, url, **kwargs) as response:
+                return response
+    except aiohttp.ClientError as err:
+        raise ExternalRequestError("Blackboard async request failed") from err
+
+
+async def _prepare_requests(method, urls, concurrency=10, **kwargs):
+    semaphore = asyncio.Semaphore(concurrency)
```

Comment:
Definitely test this to make sure but I don't think the semaphore is needed. `aiohttp.ClientSession()` has a `connector` argument which I think is a [`TCPConnector`](https://docs.aiohttp.org/en/stable/client_reference.html#tcpconnector) object by default. `TCPConnector.__init__()` has a `limit` argument which is the "total number simultaneous connections" and is `100` by default.

I don't think we need to customise this to `10`, the default of `100` is probably fine (?)